File: Main/PlatformManager.cpp
/**
 *  You can modify and use this source freely
 *  only for the development of application related Live2D.
 *
 *  (c) Live2D Inc. All rights reserved.
 */

#include <GL/glew.h>

#define STB_IMAGE_IMPLEMENTATION
#include <stb_image.h>

#include "PlatformManager.h"
#include "FileManager.h"
#include "util/UtDebug.h"
#include "LAppTextureDesc.h"
#include <Live2DModelOpenGL.h>
#include "Log.hpp"


using namespace live2d;
using namespace live2d::framework;

PlatformManager::PlatformManager(void)
{
}


PlatformManager::~PlatformManager(void)
{
}


unsigned char* PlatformManager::loadBytes(const char* path, size_t* size)
{
	unsigned char* data = FileManager::loadFile(path ,(int *)size);
	return data;
}

void PlatformManager::releaseBytes(void* data)
{
	FileManager::releaseBuffer(data);
}

ALive2DModel* PlatformManager::loadLive2DModel(const char* path)
{
	size_t size;
	unsigned char* buf = loadBytes(path, &size);
	
	//Create Live2D Model Instance
	ALive2DModel* live2DModel = Live2DModelOpenGL::loadModel(buf, (int)size);

    return live2DModel;
}

L2DTextureDesc* PlatformManager::loadTexture(ALive2DModel* model, int no, const char* path)
{
	unsigned int textureID;

	int width, height, channels;
	assert(textureCount <= 10);
	unsigned char* data = stbi_load(path, &width, &height, &channels, 0);

	assert(data != NULL);

	glEnable(GL_TEXTURE_2D);
	glGenTextures(1, &textureID);

	glBindTexture(GL_TEXTURE_2D, textureID);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);

	/*glTexImage2D(
	GL_TEXTURE_2D , 0 , GL_RGBA , width , height ,
	0 , GL_RGBA , GL_UNSIGNED_BYTE , image
	);*/
	gluBuild2DMipmaps(GL_TEXTURE_2D, GL_RGBA, width, height, GL_RGBA, GL_UNSIGNED_BYTE, data);

	stbi_image_free(data);
	LAppTextureDesc* desc=new LAppTextureDesc(textureID);

	((live2d::Live2DModelOpenGL*)model)->setTexture(no, textureID);
	return desc;
}

void PlatformManager::log(const char* txt)
{
	Info("%s", txt);
}

File: package/live2d/v2/platform_manager.py
import json
from typing import Any, Dict

import OpenGL.GL as gl
from PIL import Image

from .core import Live2DModelOpenGL

class PlatformManager:

    def loadBytes(self, path) -> bytes:
        with open(path, 'rb') as f:
            return f.read()

    def loadLive2DModel(self, path) -> Live2DModelOpenGL:
        with open(path, 'rb') as f:
            return Live2DModelOpenGL.loadModel(f.read())

    def loadTexture(self, live2DModel, no, path):
        image = Image.open(path)
        if image.mode != 'RGBA':
            image = image.convert("RGBA")
        image_data = image.tobytes()
        width, height = image.size
        gl.glEnable(gl.GL_TEXTURE_2D)
        texture = gl.glGenTextures(1)
        gl.glBindTexture(gl.GL_TEXTURE_2D, texture)
        gl.glTexImage2D(
            gl.GL_TEXTURE_2D, 0, gl.GL_RGBA, width, height, 0, gl.GL_RGBA, gl.GL_UNSIGNED_BYTE, image_data
        )
        gl.glTexParameteri(gl.GL_TEXTURE_2D, gl.GL_TEXTURE_MIN_FILTER, gl.GL_LINEAR_MIPMAP_NEAREST)
        gl.glTexParameteri(gl.GL_TEXTURE_2D, gl.GL_TEXTURE_MAG_FILTER, gl.GL_LINEAR)
        gl.glGenerateMipmap(gl.GL_TEXTURE_2D)
        gl.glBindTexture(gl.GL_TEXTURE_2D, 0)
        live2DModel.setTexture(no, texture)

    def jsonParseFromBytes(self, path) -> Dict[str, Any]:
        return json.loads(path)
